main/tests/server/src/com/google/refine/grel/controls/FilterTests.java

```diff
@@ -27,15 +27,48 @@
 
 package com.google.refine.grel.controls;
 
+import static org.testng.Assert.assertTrue;
+import static org.testng.Assert.fail;
+
+import java.util.Properties;
+
 import org.testng.annotations.Test;
 
+import com.google.refine.expr.EvalError;
+import com.google.refine.expr.Evaluable;
+import com.google.refine.expr.MetaParser;
+import com.google.refine.expr.ParsingException;
+import com.google.refine.grel.GrelTestBase;
 import com.google.refine.util.TestUtils;
 
-public class FilterTests {
+public class FilterTests extends GrelTestBase {
 
     @Test
     public void serializeFilter() {
         String json = "{\"description\":\"Evaluates expression a to an array. Then for each array element, binds its value to variable name v, evaluates expression test which should return a boolean. If the boolean is true, pushes v onto the result array.\",\"params\":\"expression a, variable v, expression test\",\"returns\":\"array\"}";
         TestUtils.isSerializedTo(new Filter(), json);
     }
+
+    private void assertParseError(String expression) throws ParsingException {
+        Evaluable eval = MetaParser.parse("grel:" + expression);
+        Object result = eval.evaluate(bindings);
+        assertTrue(result instanceof EvalError, "Expression didn't return error : " + expression);
+    }
+
+    @Test
+    public void testInvalidParams() throws ParsingException {
+        bindings = new Properties();
+        bindings.put("v", "");
+        assertParseError("filter('test', v, v)");
```

Comment:
I think this should also be wrapped in a try/throw in order for us to check both that the first argument is rejected because it's not an array, and the catch so we check that the result of that is that we throw an EvalError rather than a string?

Note this doesn't seem to have been done in forEach but I think it's the right thing to do
@wetneb am I right?

Reply:
Intuitively, rejecting the first argument because it's not array should not throw an exception at parsing stage, it should return an EvalError at evaluation stage, so there shouldn't be a need for a try/catch here.
Besides, to check that something throws an exception it's better to use `assertThrows()` as it makes the assertion clearer and the code simpler.

Reply:
No need to clean up the existing ones as I have a pull request open to address that: https://github.com/OpenRefine/OpenRefine/pull/6645. But it's helpful if you can use `assertThrows` for any new assertions you are making.